### PR TITLE
[coordinator] Validate placement on set placement endpoint unless force set

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

While this endpoint isn't mean to be extensively leveraged, it is useful for certain operations and as such it is worth validating the placement unless validation is opted out of.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/cluster/placement/placement.go [hunk 1]
```diff
@@ -311,6 +319,7 @@ func Validate(p Placement) error {
 	totalLeaving := 0
 	totalInit := 0
 	totalInitWithSourceID := 0
+	instancesLeavingShardsWithMatchingInitShards := make(map[string]map[uint32]string)
```
Comment (arnikola) [inline src/cluster/placement/placement.go:322]: nit; never uses the string value, maybe this should be a `make(map[string]map[uint32]struct{})`?

**Review by schallert — APPROVED** [review]

File: src/cluster/placement/placement_test.go [hunk 2]
```diff
@@ -276,6 +277,87 @@ func TestValidateLeavingNotMatchInitializingWithSourceID(t *testing.T) {
...
+func TestValidateLeavingAndInitializingWithSourceIDDoubleMatched(t *testing.T) {
+	i1 := NewEmptyInstance("i1", "r1", "z1", "endpoint", 1)
+	i1.Shards().Add(shard.NewShard(1).SetState(shard.Available))
+	i1.Shards().Add(shard.NewShard(2).SetState(shard.Leaving))
+	i1.Shards().Add(shard.NewShard(3).SetState(shard.Available))
+
+	i2 := NewEmptyInstance("i2", "r2", "z1", "endpoint", 1)
+	i2.Shards().Add(shard.NewShard(2).SetState(shard.Initializing).SetSourceID("i1"))
+
+	i3 := NewEmptyInstance("i3", "r2", "z1", "endpoint", 1)
```
Comment (arnikola) [inline src/cluster/placement/placement_test.go:348]: nit: repeated code may tick off the linter unfortunately, can we just add stuff to the placement options and attempt validation in one function? Validates 1 test per function/path, but probably easier to maintain and see what's going on

**Review by arnikola — APPROVED** [review]
lgtm w nits

File: src/cluster/placement/placement.go (hunk 1 above)
Comment (robskillington) [reply src/cluster/placement/placement.go:322]: We do to print out the error of which host previously had claimed that leaving shard to match it's own initializing shard:
```go
					match := matches[s.ID()]
					if match != "" {
						return fmt.Errorf(
							"instance %s has initializing shard %d with "+
								"source ID %s but leaving instance has shard already matched by %s",
							instance.ID(), s.ID(), sourceID, match)
					}
```

**Pull request merged**